### ISSUE TITLE
Additional CI/CD Improvements

### ISSUE THREAD

**Issue by CleanCut** [amethyst/amethyst#2407]
We moved from Jenkins to GitHub Actions for CI in https://github.com/amethyst/amethyst/pull/2382 -- we wanted to ship that so we can get CI working in general for folks again.  Here's a list of other things we would like to do soon.  Please comment if you have more ideas!

Each item should have individual issue(s) and/or PR(s) opened for them.  Let me know in a comment and I'll add a link next to the item you are working on (or edit this comment directly if you have sufficient privileges).

- [ ] Review webhooks and remove any that aren't used
- [ ] Make clippy warnings fail
- [ ] Use [`mdbook-linkcheck`](https://github.com/Michael-F-Bryan/mdbook-linkcheck) to detect broken book links @Blisto91 
- [ ] Use [`cargo-deadlinks`](https://github.com/deadlinks/cargo-deadlinks) to detect broken links @Blisto91 
- [ ] Add CODEOWNERS to cover things like Cargo.toml files.
- [ ] Speed up checks.  Goal: 10 minutes or less
  - [ ] use `sccache` with S3 for `cargo test`
  - [ ] use self-hosted runners
  - [ ] Experiment with various compiler settings for reducing compile times and/or sizes
  - [ ] de-duplicate dependencies
- [x] Run all examples #2500
- [ ] Automatically push rendered API docs to website (A script - #2445)
  - [ ] master merges
  - [ ] releases
  - [ ] don't include docs for dependencies
- [ ] Automatically push rendered book to website
  - [ ] master merges
  - [ ] releases
- [ ] Publish API docs to `docs.rs` instead of the web site @Blisto91 #2438 
- [ ] Automatic updating of book/doc links during release (links to stable vs master). @Blisto91 
- [ ] Check for undocumented public items
- [ ] Require a change log entry unless a specific label (or something) is present
- [ ] Find way to post comments to PRs with guidance for figuring out how to fix some CI failures. "Clippy failed, please run `cargo clippy` and fix all errors..." -- _I don't know if this is feasible_
- [ ] Find some way to indicate failures on `nightly` without making the build look like a red X.
  - You _can_ let people merge by not requiring a specific check to pass, but the build will be a big red failure even if the failing check wasn't required. That's not a good experience. ☹️
  - @Blisto91 I seem to recall you suggested an approach that used an action that commented to the PR, but I can't find the source.  Can you refresh my memory?
- [ ] How much can we automate blog posts?  Minor patches should only need the changlog posted anyway.
- [ ] Coverage
- [ ] Automate publishing crates, perhaps using [cargo hack](https://crates.io/crates/cargo-hack) as per https://github.com/amethyst/amethyst/pull/2437#issuecomment-674935448
- [x] Sanitize dependencies with [cargo-deny](https://github.com/EmbarkStudios/cargo-deny) -- see also https://github.com/amethyst/amethyst/issues/1948
- [x] Simplify the Pull Request template (people don't _have_ to manually run what CI will catch)
  - It would be great to move CI commands into a series of scripts and mention in the PR template what scripts folks could run locally to get quicker feedback

**Comment by Blisto91** [comment]
So it is possible to look at the the status of a step [before continue-on-error is applied. ](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#steps-context)
So a separate step could either post a comment depending on the outcome or set a job output and then have a separate job post a comment afterwards.
There are quite a few different actions that let you post comments on a pr. I've toyed around [with this one myself. ](https://github.com/marocchino/sticky-pull-request-comment)

This could be a workaround of allowing failure on nightly without the checks glowing red. A bot would notify of the failure in the PR without blocking anything.
Maybe a separate nightly badge in the readme could show a daily status if that would be of interest.

**Comment by CleanCut** [comment]
FYI, in general I plan to open a single PR per task on this list so I can merge more frequently.

**Comment by Blisto91** [comment]
@CleanCut i'll try yo tackle the docs.rs situation -> [#2438](https://github.com/amethyst/amethyst/issues/2438)

**Comment by Blisto91** [comment]
Another thing I would like on the list is support for automatic update of book/doc links during release.

Currently the API links in the book are a mix of stable and master no matter the version linked from. 
The documentation should in general only link to the same version. Stable links to stable and master links to master.
This should help avoid dead or outdated links.

The book/docs on the github repo should only link to master (except maybe some special cases) and then during a release a script could update the relevant links to point to a stable link.

**Comment by Veetaha** [comment]
Hi guys, I have an experience setting up ci for Rust in my company.
Here are some of my tips and suggestions:
- We currently use a setup of `sccache` with Redis (because of security concerns with s3). I'd be glad to handle this for `amethyst` :D
- Regarding making clippy warning fail: we just have `RUSTC_FLAGS: RUSTFLAGS: --deny warnings -Cdebuginfo=0`
- Regarding `-Cdebuginfo=0` on CI: we found that Github-hosted runners have too small disk space and with our initial migration to Github actions we encountered disk space overflow. I have optimized our disk usage by disabling debug info and moving our integration tests to use single-binary setup (which was the most impactful change, I guess). More details that in the issue I originally stole the knowledge about this: https://github.com/rust-lang/cargo/issues/4867 and an its implementation PR: https://github.com/rust-lang/cargo/pull/5022#issuecomment-364691154
- Looks like you already have `mdbook-linkcheck` on ci, but commented out cc @CleanCut?
https://github.com/amethyst/amethyst/blob/965e2c4bba5f8751fc69dfe74583be8332b0ebbd/.github/workflows/ci.yml#L91-L97
- Regarding sending comments to PR: this is possible by using just Github API with curl with [`secrets.GITHUB_TOKEN`](https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#using-the-github_token-in-a-workflow) or a [github action](https://github.com/unsplash/comment-on-pr), but the action adds additional overhead and issues (like rubygems unavailability).
- In general, we have concluded that sending comments to PR should be set to a minimum. Sending a comment due to clippy check failure looks unnecessary to me, it won't bring too much value. When ci fails we generally just open the action run logs and figure out what happened, inspect clippy logs and fix our code. I wouldn't even go for sending notifications to PR when nightly builds fail, because red cross with [`continue-on-error: true`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error) is enough for people to notice, I think.
- Regarding coverage: unfortunately, Rust seems to be quite immature in this field. We use [`grcov `](https://github.com/mozilla/grcov) with [`codecov.io`](https://codecov.io/) on our CI but this job requires `nightly` toolchain and the coverage reports are quite unreliable, enough to mention that `grcov` reports different coverage stats for a PR that doesn't even change Rust code (seems ridiculous, huh...). However, I have been looking at [cargo-tarpaulin](https://crates.io/crates/cargo-tarpaulin) lately, I am not sure whether it is better than `grcov`, but I would like to try this out at some point of time.
- I would suggest adding [cargo-deny](https://github.com/EmbarkStudios/cargo-deny) check to the list (we use it in our open-sourced repo [here](https://github.com/elastio/devx/blob/d8b39f41c8790f0616d01098025cb962bc63c675/.github/workflows/ci.yml#L23-L33))
It has been a pleasure to use this tool to sanitize our dependencies, detect multiple entries of different crate versions with a convenient [configuration file](https://github.com/elastio/devx/blob/d8b39f41c8790f0616d01098025cb962bc63c675/deny.toml#L1-L39). It also looks like `rustsec` is tending towards making this tool its official frontend: https://github.com/EmbarkStudios/cargo-deny/issues/194
- Not sure if `amethyst` really needs `cargo-hack`, I would avoid dependencies unless really necessary. It is a known fact that `cargo` doesn't support publishing the whole workspace via a subcommand, so we have to manually traverse the crates and publish them starting from leaf packages to the root ones. Here is an example of [a job](https://github.com/elastio/devx/blob/d8b39f41c8790f0616d01098025cb962bc63c675/.github/workflows/ci.yml#L73-L102) that publishes the crates when a new git tag is pushed to the repo.
- For ci performance sake it might make sense to break up `cargo clippy` and `cargo fmt --check`, `cargo test` and other checks into multiple ci jobs so that they can run in parallel. However, note that this will be bounded by the fact that both of these jobs will need to compile the workspace on their own, but it looks like even so, parallel checks seem to be faster (though I haven't measured :D).

#### Questions
- How does the process of creating VMs and S3 buckets for `amethyst` hosted runners and `sccache` goes? Who is the owner of these assets and how would it influence the task for integration `sccache` and related stuff?

#### Some other points not yet mentioned or brought up
##### Caching `CARGO_HOME`:
We had some problems with caching [`CARGO_HOME` subdirs](https://doc.rust-lang.org/cargo/guide/cargo-home.html#caching-the-cargo-home-in-ci). We haven't resolved them and abandoned chaching `CARGO_HOME` dirs at all in part due to some internal reorganization of our hosted runners, but we encountered huge cache sizes and restoration times, here is an example restoring `~/.cargo/registry/index` cache for 7 minutes...:
![image](https://user-images.githubusercontent.com/36276403/90618201-51f00b80-e218-11ea-8d33-6ce5b8cc48fb.png)
My gut feeling is that this is somehow connected with the fact that we reused the cache key across multiple linux distros, but that's just a wild guess...

**Comment by Blisto91** [comment]
@Veetaha in regards to restore times, they should have been improved with the 2.1 release of github cache. 
Tho I have still seen a fresh cache save take veeeery long.

I can see that you are using cache v1 in that picture. V2 uses zstd compression (except on windows due to a bug) and I've seen impressive size reductions with it. It's also pretty fast at decompressing. 
I've had an okay experience caching the dependencies in the cargo home dir. But it's not a huge time saving on It's own with also tackling the target folder.
Also a hash key made on linux or Mac doesn't trigger on windows due to different compression, so it can't be shared between all targets currently sadly. 

Cargo hack is only being looked at for when a new release is being published to crates.io. To avoid having to manually comment out the dev-dependencies. 

Think it would be great with help from someone with practical experience! ♥

**Comment by CleanCut** [comment]
> Another thing I would like on the list is support for automatic updating of book/doc links during release.

✅ Added to the list.

**Comment by CleanCut** [comment]
> Hi guys, I have experience setting up ci for Rust in my company.

@Veetaha Awesome! We can use all the help we can get.

>     * We currently use a setup of `sccache` with Redis (because of security concerns with s3). I'd be glad to handle this for `amethyst` :D

Great! If you will open a PR with the change using environment variables wherever you need AWS stuff (url, secrets, etc.), and then I'll follow up on the AWS side and get the secrets set up for CI.  I'd actually like to try both Redis _and_ s3 so we can compare price/performance, but we can start here.

>     * Looks like you already have `mdbook-linkcheck` on ci, but commented out cc @CleanCut?

Yes, because there are existing broken links that need to be fixed. If you'd like to open a PR to fix the broken links and enable this code, I'd be happy to review.

>     * I would suggest adding [cargo-deny](https://github.com/EmbarkStudios/cargo-deny) check to the list

✅ Added to the list.

>     * For ci performance sake it might make sense to break up...

This saved very little time when I tried it. 😢  We need to be careful with how many total builds we do.  We want to use fast self-hosted runners with more CPU and RAM, but we have to pay for their usage!  If we have too many builds, it will cost too much and we will be forced to remain on shared runners.  If we end up staying with shared runners in the end, then I agree that we should try breaking things up as much as we can to parallelize the builds.

>     * How does the process of creating VMs and S3 buckets for `amethyst` hosted runners and `sccache` goes? Who is the owner of these assets and how would it influence the task for integrating `sccache` and related stuff?

The Amethyst Foundation (a non-profit) owns the AWS account, etc.  Right now, the process is I ask leadership for access when I have a direct need.  If the cost of a solution is too high, we'll have to shut it down and not use it even if it works well.

>     * Looks like this file for gitlab ci isn't used https://github.com/amethyst/amethyst/blob/master/gitlab-ci-helper.sh?

It is not used in CI, _yet_.  That's what I used to manually generate the docs for the 0.15.1 release (and I do mean manually - I picked out specific commands from the script and ran them myself). It's on the list in several places already to turn this into actual automation. 😄 I've already renamed it and started working on it a bit in #2437 and will soon be opening another PR to finish it up.

---

For all the things I didn't respond to directly: Feel free to open a PR and ping me!

**Comment by Veetaha** [comment]
I think Redis will be much more expensive than just S3, because for Redis you'd need to run a VM. We use [`Hetzner`](https://www.hetzner.com/cloud?pk_campaign=8482717093&pk_source=google&pk_keyword=cloud%20server&pk_medium=cpc&pk_content=404849758926&gclid=CjwKCAjwm_P5BRAhEiwAwRzSOyyXTKzkHC-h6pHiMJ6VU9czse1mczEpALQI8csoIPbZcL_j51fBhxoCuUMQAvD_BwE) for running our self-hosted runners in vagrant and our Redis instance too. I am not the one who came up with using this provider, but it provides baremetal machines for reasonably low cost. But this is just FYI

**Comment by Blisto91** [comment]
@CleanCut You can sign me up for fixing book links and getting mdbook-linkcheck running.

**Comment by Blisto91** [comment]
In relation [to this error](https://github.com/amethyst/amethyst/pull/2455#issuecomment-687863739).

I was trying to find a way of preventing the other jobs from cancelling when nightly fails. But it seems you cannot set fail-fast to false on a specific matrix job.
You can do it with continue-on-error but that includes the downsides of that setting.
You also currently can't prevent a specific matrix job from running depending on the branch (bors staging or trying).
But we can do it if we split nightly into a separate job.

I currently see 4 (5) ways around this bors issue.

1. Continue-on-error true on nightly.
2. Move nightly to separate job not in a matrix and only run when not in bors branch.
3. Disable fail-fast on the whole matrix.
4. Remove nightly completely from CI
5? Make bors run a separate CI that is a copy of the main one but without nightly. Dunno if this is possible.

**Comment by CleanCut** [comment]
@Blisto91 Strategy number 1 seems to have worked for me in https://github.com/amethyst/amethyst/pull/2455 ... so that's our current de-facto strategy since it got merged.  Feel free to change it to something better if you like! 😄

**Comment by ezpuzz** [comment]
cargo-deny check implemented in #2590
